Sources/HummingbirdWebSocket/Server/WebSocketRouter.swift

```diff
@@ -0,0 +1,157 @@
+        _ path: String = "",
+        shouldUpgrade: @Sendable @escaping (Request, Context) async throws -> RouterShouldUpgrade = { _, _ in .upgrade([:]) },
+        handle: @escaping WebSocketDataCallbackHandler.Callback
+    ) -> Self where Context: WebSocketRequestContext {
+        return on(path, method: .get) { request, context -> Response in
+            let result = try await shouldUpgrade(request, context)
+            switch result {
+            case .dontUpgrade:
+                return .init(status: .notAcceptable)
+            case .upgrade(let headers):
+                context.webSocket.handler.withLockedValue { $0 = WebSocketDataCallbackHandler(handle) }
+                return .init(status: .ok, headers: headers)
+            }
+        }
+    }
+}
+
+extension HTTP1AndWebSocketChannel {
+    ///  Initialize HTTP1AndWebSocketChannel with async `shouldUpgrade` function
+    /// - Parameters:
+    ///   - additionalChannelHandlers: Additional channel handlers to add
+    ///   - responder: HTTP responder
+    ///   - maxFrameSize: Max frame size WebSocket will allow
+    ///   - webSocketRouter: WebSocket router
+    /// - Returns: Upgrade result future
+    public init<Context: WebSocketRequestContext, ResponderBuilder: HTTPResponderBuilder>(
+        additionalChannelHandlers: @escaping @Sendable () -> [any RemovableChannelHandler] = { [] },
+        responder: @escaping @Sendable (Request, Channel) async throws -> Response = { _, _ in throw HTTPError(.notImplemented) },
+        maxFrameSize: Int = (1 << 14),
+        webSocketRouter: ResponderBuilder
+    ) where Handler == WebSocketDataCallbackHandler, ResponderBuilder.Responder.Context == Context {
+        let webSocketRouterResponder = webSocketRouter.buildResponder()
+        self.init(additionalChannelHandlers: additionalChannelHandlers, responder: responder, maxFrameSize: maxFrameSize) { head, channel, logger in
+            let request = Request(head: head, body: .init(buffer: .init()))
+            let context = Context(channel: channel, logger: logger.with(metadataKey: "hb_id", value: .stringConvertible(RequestID())))
+            do {
+                let response = try await webSocketRouterResponder.respond(to: request, context: context)
+                if response.status == .ok, let webSocketHandler = context.webSocket.handler.withLockedValue({ $0 }) {
+                    return .upgrade(response.headers, webSocketHandler)
+                } else {
+                    return .dontUpgrade
+                }
+            } catch {
+                return .dontUpgrade
+            }
+        }
+    }
+}
+
+extension HTTPChannelBuilder {
+    /// HTTP1 channel builder supporting a websocket upgrade
+    ///  - parameters
+    public static func webSocketUpgrade<ResponderBuilder: HTTPResponderBuilder>(
+        additionalChannelHandlers: @autoclosure @escaping @Sendable () -> [any RemovableChannelHandler] = [],
+        maxFrameSize: Int = 1 << 14,
```

Comment:
Can we make this a parameter `configuration: WebSocketServerConfiguration`?
`maxFrameSize` is definitely the most important configurable property, but I can imagine us adding a few more in the future.